ark/phenotyping/cluster.py

```diff
@@ -0,0 +1,151 @@
+import os
+
+import numpy as np
+import pandas as pd
+
+
+def winner(sample, weights):
+    """Find the coordinates to the winning neuron in the SOM
+
+    Args:
+        sample (numpy.ndarray):
+            A row in the pixel matrix identifying information associated with one pixel
+        weights (numpy.ndarray):
+            A square matrix identifying the weights between two neurons
+
+    Returns:
+        tuple:
+            A coordinate array which indicates the winning neuron's position
+    """
+
+    # get euclidean distance between the sample and the weights
+    activation_map = np.linalg.norm(np.subtract(sample, weights), axis=-1)
+
+    # find the winning neuron's coordinates, needed for easy access
+    winning_coords = np.unravel_index(activation_map.argmin(), activation_map.shape)
+
+    return winning_coords
+
+
+def update(sample, weights, winning_coords, sigma, learning_rate,
+           x_mesh, y_mesh, t, num_iters):
+    """Updates the weights, learning rate, and sigma parameters
+
+    Args:
+        sample (numpy.ndarray):
+            A row in the pixel matrix identifying information associated with one pixel
+        weights (numpy.ndarray):
+            A square matrix identifying the weights between two neurons
+        winning_coords (tuple):
+            A coordinate array which indicates the winning neuron's position
+        sigma (float):
+            Determines the spread of the Gaussian neighborhood function
+        learning_rate (float):
+            Determines how sensitive the weight updates will be to new data
+        x_mesh (numpy.ndarray):
+            The x coordinate matrix of the weights vectors
+        y_mesh (numpy.ndarray):
+            The y coordinate matrix of the weights vectors
+        t (int):
+            The current iteration index
+        num_iters (int):
+            The maximum number of iterations
+
+    Returns:
+        numpy.ndarray:
+            The updated weights matrix
+    """
+
+    # update learning rate with asymptotic decay
+    decay_lr = learning_rate / (1 + t / (num_iters / 2))
+
+    # update sigma with asymptotic decay
+    decay_sig = sigma / (1 + t / (num_iters / 2))
+
+    # return a Gaussian centered around the winning coordinates
+    d = 2 * np.pi * decay_sig * decay_sig
+    ax = np.exp(-np.power(x_mesh - x_mesh.T[winning_coords], 2) / d)
+    ay = np.exp(-np.power(y_mesh - y_mesh.T[winning_coords], 2) / d)
+    g = (ax * ay).T * decay_lr
+
+    # update the weights based on the Gaussian neighborhood
+    weights += np.einsum('ij, ijk->ijk', g, sample - weights)
+
+    return weights
+
+
+def train_flowsom(pixel_mat, x_neurons, y_neurons, num_iters,
```

Comment:
FlowSOM specifically refers to the R package that combines SOM with minimum spanning tree and some nice visualization. We can just use SOM in our function names